Cython/Compiler/Nodes.py

```diff
@@ -2190,7 +2175,17 @@ def generate_function_definitions(self, env, code):
             if entry.type.is_buffer:
                 Buffer.put_acquire_arg_buffer(entry, code, self.pos)
 
-        if acquire_gil_for_var_decls_only:
+        if code.funcstate.needs_refnanny:
+            # if this is true there's definite some reference counting in
+            # the variable declarations
+            var_decls_need_gil = True
+
+        if var_decls_need_gil and lenv.nogil and gilstate_decl is not None:
+            gilstate_decl.put_ensure_gil()
+            code.funcstate.gil_owned = True
+            gilstate_decl = None
+
+        if var_decls_need_gil and lenv.nogil:
             code.put_release_ensured_gil()
             code.funcstate.gil_owned = False
```

Comment:
I think this is the same as this:
```suggestion
        if var_decls_need_gil and lenv.nogil:
            if gilstate_decl is not None:
                gilstate_decl.put_ensure_gil()
                gilstate_decl = None
                code.funcstate.gil_owned = True
            code.put_release_ensured_gil()
            code.funcstate.gil_owned = False
```